lib/eslint/eslint-helpers.js

```diff
@@ -151,94 +285,183 @@ async function globSearch({ basePath, patterns, configs }) {
+        ([basePath, { patterns, rawPatterns }]) => ({ basePath, patterns, rawPatterns })
+    ).filter(({ patterns }) => patterns.length > 0);
+
+    const results = await Promise.allSettled(
+        normalizedSearches.map(
+            ({ basePath, patterns, rawPatterns }) => globSearch({
+                basePath,
+                patterns,
+                rawPatterns,
+                configs,
+                errorOnUnmatchedPattern
+            })
         )
     );
 
-    return [...new Set(results.flat())];
-}
-
-/**
- * Determines if a given glob pattern will return any results.
- * Used primarily to help with useful error messages.
- * @param {Object} options The options for the function.
- * @param {string} options.basePath The directory to search.
- * @param {string} options.pattern A glob pattern to match.
- * @returns {Promise<boolean>} True if there is a glob match, false if not.
- */
-function globMatch({ basePath, pattern }) {
-
-    let found = false;
-    const patternToUse = path.isAbsolute(pattern)
-        ? normalizeToPosix(path.relative(basePath, pattern))
-        : pattern;
+    const filePaths = [];
 
-    const matcher = new Minimatch(patternToUse);
+    for (let i = 0; i < results.length; i++) {
 
-    const fsWalkSettings = {
+        const result = results[i];
+        const currentSearch = normalizedSearches[i];
 
-        deepFilter(entry) {
-            const relativePath = normalizeToPosix(path.relative(basePath, entry.path));
+        if (result.status === "fulfilled") {
 
-            return !found && matcher.match(relativePath, true);
-        },
-
-        entryFilter(entry) {
-            if (found || entry.dirent.isDirectory()) {
-                return false;
+            // if the search was successful just add the results
+            if (result.value.length > 0) {
+                filePaths.push(...result.value);
+                continue;
             }
 
-            const relativePath = normalizeToPosix(path.relative(basePath, entry.path));
-
-            if (matcher.match(relativePath)) {
-                found = true;
-                return true;
+            // if the are no results we need to check for ignored files
+            if (errorOnUnmatchedPattern) {
+                await checkForIgnoredResults(currentSearch);
             }
 
-            return false;
+            continue;
         }
-    };
 
-    return new Promise(resolve => {
+        if (errorOnUnmatchedPattern) {
 
-        // using a stream so we can exit early because we just need one match
-        const globStream = fswalk.walkStream(basePath, fsWalkSettings);
+            // if we make it here then there was an error
+            const error = result.reason;
 
-        globStream.on("data", () => {
-            globStream.destroy();
-            resolve(true);
-        });
+            // unexpected errors should be re-thrown
+            if (!error.basePath) {
+                throw error;
+            }
```

Comment:
I think this should be outside `if (errorOnUnmatchedPattern) {}`, so that `errorOnUnmatchedPattern: false` does not suppress unexpected errors like file system errors. I denied permissions on `tools/internal-testers` directory, and then ran `eslint tools -f tap --no-error-on-unmatched-pattern`. It exits with 0 without linting anything. In eslintrc mode, the same command throws EPERM, which is I think better than silently ignoring the whole pattern.

Reply:
Ah good call.